### PR TITLE
question 2 completed

### PR DESCRIPTION
**Pull request by TomMirfin**
Question 2 pull req

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+const { handleCustomErrors } = require("./errors");
+
+const app = express();
+
+app.use(express.json());
+
+app.get("/api/topics", getAllTopics);
+
+app.get("/api/nvnewbs", getAllTopics);
+
```
Comment (2202Hannah) [inline app.js:12]: You should also include middleware that would capture any unexpected i.e. 500 errors

File: db/seeds/run-seed.js [hunk 2]
```diff
@@ -1,7 +1,7 @@
-const devData = require('../data/development-data/index.js');
-const seed = require('./seed.js');
-const db = require('../connection.js');
+const devData = require("../data/development-data/index.js");
+const seed = require("./seed.js");
 
+const db = require("./");
```
Comment (2202Hannah) [inline db/seeds/run-seed.js:4]: Why did you change the path of your database connection file? Does this still work?

**Review by 2202Hannah — CHANGES_REQUESTED** [review]
Hi Tom, really nice work on this ticket so far. Your test suite is set up well and you have made sure to seed the database before each test. You are also using the MVC structure which is great to see.

There seems to be some confusion when you are handling an unexpected endpoint/ a typo. Please consider my in line comments and update your code - I would like to see you using app.all to handle an unexpected endpoint rather than the conditional logic you have implemented in the controller. 

Please submit another pr when you have made the changes. Any questions then please give us a shout on nchelp :)

File: __tests__/endpoint.test.js [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  test("status:404, responds with an error message when passed a bad path", () => {
+    return request(app)
+      .get("/api/nvnewbs")
+      .expect(404)
+      .then(({ body }) => {
+        expect(body.msg).toBe("Not Found");
+      });
+  });
+});
+// describe("/api", () => {
```
Comment (itskatherine) [inline __tests__/endpoint.test.js:46]: Please get rid of commented out code 🗑️

File: app.js [hunk 4]
```diff
@@ -1,13 +1,23 @@
...
 const app = express();
 
 app.use(express.json());
 
 app.get("/api/topics", getAllTopics);
 
-app.get("/api/nvnewbs", getAllTopics);
+app.get("/api", (req, res) => {
+  // console.log(listEndpoints(app));
+  res.status(200).send(listEndpoints(app));
```
Comment (itskatherine) [inline app.js:15]: This should be on the next ticket not this one

File: app.js [hunk 5]
```diff
@@ -1,13 +1,23 @@
...
+});
+
+app.use(handleCustomError);
+app.use(handleServerErrors);
+
+app.all("/*", getAllTopics);
```
Comment (itskatherine) [inline app.js:21]: This is invoking `getAllTopics` when the user puts in an incorrect endpoint. 
Instead, you should have a separate function for handling this, which sends off a 404 and an error message

File: errors.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+exports.handleCustomError = (req, res, err) => {
+  console.log(err.code, "<--- err code");
+  if (err) {
+    res.status(404).send({ msg: "Not Found" });
+  }
+  console.log(err, "<-- err");
+};
+
+exports.handleServerErrors = (err, req, res, next) => {
```
Comment (itskatherine) [inline errors.js:9]: Nice implementation of a function to handle server errors 👍

File: errors.js [hunk 7]
```diff
@@ -0,0 +1,12 @@
+exports.handleCustomError = (req, res, err) => {
```
Comment (itskatherine) [inline errors.js:1]: We currently don't have any custom errors being implemented so we don't need this function yet

File: app.js (hunk 1 above)
Comment (itskatherine) [reply app.js:22]: You've done this ✅

File: app.js [hunk 8]
```diff
@@ -0,0 +1,13 @@
+const express = require("express");
+const { getAllTopics } = require("./controller/topics.controller");
+const { handleCustomErrors } = require("./errors");
+
+const app = express();
+
+app.use(express.json());
```
[the remark this answers is not in the document]
Comment (itskatherine) [reply app.js:9]: Still need to get rid of this line (`app.use(express.json())`)

**Review by itskatherine — COMMENTED** [review]
Hi Tom, could you please have another look at my comments and then you can merge once you've addressed them
